src/Utils.ts

```diff
@@ -986,12 +986,129 @@ function setDummyInputDebugValue(
     );
 }
 
+export class DummyInputObserver implements Types.DummyInputObserver {
+    private _win?: GetWindow;
+    private _offsetsQueue: (() => void)[] = [];
+    private _offsetsTimer?: number;
+    private _domChangedMap: WeakMap<HTMLElement, true> = new WeakMap();
+    private _domChangedTimer?: number;
+    private _dummyMap: WeakMap<HTMLElement, () => void> = new WeakMap();
+    private _dummies: WeakRef<HTMLElement>[] = [];
```

Comment:
`add`/`remove` is when the particular dummy input is added to or removed from the DOM (happens multiple times during the application life cycle). `dispose` is for the entire Tabster dispose.

The entries need to be enumerated during the update. The map is to have a hash with the element as the key (JS only supports strings as keys for normal objects) for fast checks. The list is to be able to go through them (as WeakMap doesn't support enumeration). WeakRefs are everywhere in Tabster by design, because Tabster only follows the application DOM, not creates it (and bizarre things happen in the applications) — last thing you want is to be constantly chased by the people investigating memory leaks and blaming you because you hold a strong reference.

Other than that — I just think it's a good idea to make sure things are explicitly clear on dispose.

Reply:
> The list is to be able to go through them (as WeakMap doesn't support enumeration)

If a regular Map is used, then you can iterate through keys there.

> I just think it's a good idea to make sure things are explicitly clear on dispose.

My suggestion does not change this

> last thing you want is to be constantly chased by the people investigating memory leaks and blaming you because you hold a strong reference

From what I see here is that the contract to remove dummy inputs `remove`, if something else happens to remove the reference that's actually bug, not  a feature.

Reply:
Yeah that makes sense to me. We control these HTMLElements, if there is a memory leak that's an indicator that _**we have a bug**_